### PR TITLE
Add ExecutionProgress() method to function interface to report exec progress

### PR DESCRIPTION
**Pull request by PrasadG193**
## Change Overview

This PR adds ExecProgress() method to Kanister function interface to report the function exec progress.
This PR replaces the weight based progress calculation logic with function-specific implementation.
By default, all functions report progressPercent as 0 during exec and 100 once the execution is completed.
Ideally, each function should have a way to get stats of execution progress. Which can be added as a followup 

The design has been discussed here: https://github.com/kanisterio/kanister/pull/1875

## Pull request type

Please check the type of change your PR introduces:
- [ ] :construction: Work in Progress
- [ ] :rainbow: Refactoring (no functional changes, no api changes)
- [ ] :hamster: Trivial/Minor
- [ ] :bug: Bugfix
- [ ] :sunflower: Feature
- [ ] :world_map: Documentation
- [ ] :robot: Test

## Issues <!-- to auto-close the issue, add the "fixes" keyword -->

- fixes #issue-number

## Test Plan

<!-- Will run prior to merging.-->
<!-- Include example how to run.-->

- [ ] :muscle: Manual
- [ ] :zap: Unit test
- [ ] :green_heart: E2E

### Sample output of actionSet status of each phase completion status

#### TC:1 Blueprint action consists of 3 phases and all phases succeeds

**1. First phase completed**

```
status:
  actions:
  - artifacts:
      operatorBackup:
        keyValue:
          path: '{{ .Phases.phase3.Output.path }}'
    blueprint: operator-blueprint
    deferPhase:
      name: ""
      progress: {}
      state: ""
    name: backup
    object:
      apiVersion: ""
      group: ""
      kind: statefulset
      name: mysql
      namespace: mysql
      resource: ""
    phases:
    - name: phase1
      output:
        path: /operator-backups/2023-04-19T18-42-49/backup.tgz
      progress:
        lastTransitionTime: "2023-04-19T18:43:10Z"
        progressPercent: "100"
      state: complete
    - name: phase2
      progress: {}
      state: running
    - name: phase3
      progress: {}
      state: pending
  error:
    message: ""
  progress:
    lastTransitionTime: "2023-04-19T18:43:10Z"
    percentCompleted: "33"
    runningPhase: phase2
  state: running
```

**2. Second phase completed**

```
status:
  actions:
  - artifacts:
      operatorBackup:
        keyValue:
          path: '{{ .Phases.phase3.Output.path }}'
    blueprint: operator-blueprint
    deferPhase:
      name: ""
      progress: {}
      state: ""
    name: backup
    object:
      apiVersion: ""
      group: ""
      kind: statefulset
      name: mysql
      namespace: mysql
      resource: ""
    phases:
    - name: phase1
      output:
        path: /operator-backups/2023-04-19T18-42-49/backup.tgz
      progress:
        lastTransitionTime: "2023-04-19T18:43:10Z"
        progressPercent: "100"
      state: complete
    - name: phase2
      output:
        path: /operator-backups/2023-04-19T18-42-49/backup.tgz
      progress:
        lastTransitionTime: "2023-04-19T18:43:28Z"
        progressPercent: "100"
      state: complete
    - name: phase3
      progress: {}
      state: running
  error:
    message: ""
  progress:
    lastTransitionTime: "2023-04-19T18:43:28Z"
    percentCompleted: "66"
    runningPhase: phase3
  state: running
```

**3. Third phase completed**

```
status:
  actions:
  - artifacts:
      operatorBackup:
        keyValue:
          path: /operator-backups/2023-04-19T18-42-49/backup.tgz
    blueprint: operator-blueprint
    deferPhase:
      name: ""
      progress: {}
      state: ""
    name: backup
    object:
      apiVersion: ""
      group: ""
      kind: statefulset
      name: mysql
      namespace: mysql
      resource: ""
    phases:
    - name: phase1
      output:
        path: /operator-backups/2023-04-19T18-42-49/backup.tgz
      progress:
        lastTransitionTime: "2023-04-19T18:43:10Z"
        progressPercent: "100"
      state: complete
    - name: phase2
      output:
        path: /operator-backups/2023-04-19T18-42-49/backup.tgz
      progress:
        lastTransitionTime: "2023-04-19T18:43:28Z"
        progressPercent: "100"
      state: complete
    - name: phase3
      output:
        path: /operator-backups/2023-04-19T18-42-49/backup.tgz
      progress:
        lastTransitionTime: "2023-04-19T18:43:48Z"
        progressPercent: "100"
      state: complete
  error:
    message: ""
  progress:
    lastTransitionTime: "2023-04-19T18:43:48Z"
    percentCompleted: "100"
  state: complete
```
---

#### TC:2 Blueprint consists of 3 phases, first phases succeeds and 2nd one fails

**1. First Phase started**
```
status:                                                                                                                                               
  actions:                                                                                                                                            
  - artifacts:                                                                                                                                        
      operatorBackup:                                                                                                                                 
        keyValue:                                                                                                                                     
          path: '{{ .Phases.phase3.Output.path }}'                                                                                                    
    blueprint: operator-blueprint                                                                                                                     
    deferPhase:                                                                                                                                       
      name: ""                                                             
      progress: {}                                                                                                                                    
      state: ""                                                            
    name: backup                                                                                                                                      
    object:                                                                
      apiVersion: v1                                                                                                                                  
      group: apps                                                          
      kind: statefulset                                                                                                                               
      name: mysql                                                          
      namespace: mysql                                                                                                                                
      resource: statefulsets                                                                                                                          
    phases:                                                                                                                                           
    - name: phase1                                                                                                                                    
      progress: {}                                                                                                                                    
      state: running                                                                                                                                  
    - name: phase2                                                                                                                                    
      progress: {}                                                                                                                                    
      state: pending                                                                                                                                  
    - name: phase3                                                                                                                                    
      progress: {}                                                                                                                                    
      state: pending                                                       
  error:                                                                                                                                              
    message: ""                                                            
  progress:                                                                                                                                           
    runningPhase: phase1                                                   
  state: running
```

**2. First phase completed and Second phase started**

```
status:                                                                                                                                               
  actions:                                                                                                                                            
  - artifacts:                                                                                                                                        
      operatorBackup:                                                                                                                                 
        keyValue:                                                                                                                                     
          path: '{{ .Phases.phase3.Output.path }}'                                                                                                    
    blueprint: operator-blueprint                                                                                                                     
    deferPhase:                                                                                                                                       
      name: ""                                                             
      progress: {}                                                                                                                                    
      state: ""                                                            
    name: backup                                                                                                                                      
    object:                                                                
      apiVersion: v1                                                                                                                                  
      group: apps                                                          
      kind: statefulset                                                                                                                               
      name: mysql                                                          
      namespace: mysql                                                                                                                                
      resource: statefulsets                                                                                                                          
    phases:                                                                                                                                           
    - name: phase1                                                                                                                                    
      progress: {}                                                                                                                                    
      state: running                                                                                                                                  
    - name: phase2                                                                                                                                    
      progress: {}                                                                                                                                    
      state: pending                                                                                                                                  
    - name: phase3                                                                                                                                    
      progress: {}                                                                                                                                    
      state: pending                                                       
  error:                                                                                                                                              
    message: ""                                                            
  progress:                                                                                                                                           
    runningPhase: phase1                                                   
  state: running
```

**3. Second phase failed**

```
status:                                                                                                                                                                                                     
  actions:                                                                                                                                                                                                            
  - artifacts:                                                                                                                                                                                                        
      operatorBackup:                                                                                                                                                                                                 
        keyValue:                                                                                                                                                                                                     
          path: '{{ .Phases.phase3.Output.path }}'                                                                                                                                                                    
    blueprint: operator-blueprint                                                                                                                                                                                     
    deferPhase:                                                                                                                                                                                                       
      name: ""                                                                                                                                                                                                        
      progress: {}                                                                                                                                                                                                    
      state: ""                                                                                                                                                                                                       
    name: backup                                                                                                                                                                                                      
    object:                                                                                                                                                                                                           
      apiVersion: v1                                                                                                                                                                                                  
      group: apps                                                                                                                                                                                                     
      kind: statefulset                                                                                                                                                                                               
      name: mysql                                                                                                                                                                                                     
      namespace: mysql                                                                                                                                                                                                
      resource: statefulsets                                                                                                                                                                                          
    phases:                                                                                                                                                                                                           
    - name: phase1                                                                                                                                                                                                    
      output:                                                                                                                                                                                                         
        path: /operator-backups/2023-05-22T07-32-36/backup.tgz                                                                                                                                                        
      progress:                                                                                                                                                                                                       
        lastTransitionTime: "2023-05-22T07:32:54Z"                                                                                                                                                                    
        progressPercent: "100"                                                                                                                                                                                         
      state: complete                                                                                                                                                                                                 
    - name: phase2                                                                                                                                                                                                    
      progress: {}                                                                                                                                                                                                    
      state: failed                                                                                                                                                                                                   
    - name: phase3                                                                                                                                                                                                    
      progress: {}                                                                                                                                                                                                    
      state: pending                                                                                                                                                                                                  
  error:                                                                                                                                                                                                              
    message: "Failed while waiting for Pod kanister-job-sk2bk to complete: Pod failed                                                                                                                                 
      or did not transition into complete state: Running backup\
     .
     .
  progress:                                                                                                                                                             
    lastTransitionTime: "2023-05-22T07:32:54Z"                                          
    percentCompleted: "33"                                                           
  state: failed                                                                     
```
---

#### TC:3 Blueprint with actions having multiple phases. And actionset running multiple action

**1. First phase if each action is running**

```jsx
status:       
  actions:        
  - artifacts: 
      operatorBackup:
        keyValue:
          path: '{{ .Phases.backupPhase3.Output.path }}'
    blueprint: operator-blueprint
    deferPhase:        
      name: ""   
      progress: {}    
      state: ""   
    name: backup
    object:              
      apiVersion: ""
      group: ""     
      kind: statefulset  
      name: mysql 
      namespace: mysql
      resource: ""       
    phases:       
    - name: backupPhase1
      progress:
        lastTransitionTime: "2023-08-11T06:41:08Z"
        progressPercent: "0"
      state: running           
    - name: backupPhase2
      progress: {}
      state: pending               
    - name: backupPhase3
      progress: {}
      state: pending                             
  - blueprint: operator-blueprint                 
    deferPhase:                                    
      name: ""                                     
      progress: {}                                   
      state: ""                                       
    name: restore                                      
    object:                                              
      apiVersion: ""                                      
      group: ""                                           
      kind: statefulset                                   
      name: mysql                                         
      namespace: mysql                                    
      resource: ""                                        
    phases:                                               
    - name: restorePhase1                                 
      progress: {}                                        
      state: running                                      
    - name: restorePhase2                                 
      progress: {}                                        
      state: pending                                      
    - name: restorePhase3                                 
      progress: {}                                        
      state: pending                                      
  error:                                                  
    message: ""                                           
  progress:                                               
    lastTransitionTime: "2023-08-11T06:41:08Z"            
    percentCompleted: "0"                                 
    runningPhase: restorePhase1                           
  state: running
```

**2. First phase of second action is completed**

```jsx
status:                                                                                                                                                                                                                                 
  actions:                                                
  - artifacts:                                            
      operatorBackup:                                     
        keyValue:                                         
          path: '{{ .Phases.backupPhase3.Output.path }}'                                                            
    blueprint: operator-blueprint                         
    deferPhase:                                           
      name: ""                                            
      progress: {}                                        
      state: ""                                           
    name: backup                                          
    object:                                               
      apiVersion: ""                                      
      group: ""                                           
      kind: statefulset                                   
      name: mysql                                         
      namespace: mysql                                    
      resource: ""                                        
    phases:                                               
    - name: backupPhase1                                  
      progress:                                           
        lastTransitionTime: "2023-08-11T06:41:08Z"                                                                  
        progressPercent: "0"                              
      state: running                                      
    - name: backupPhase2                                  
      progress: {}                                        
      state: pending                                      
    - name: backupPhase3                                  
      progress: {}                                        
      state: pending                                      
  - blueprint: operator-blueprint                         
    deferPhase:                                           
      name: ""                                            
      progress: {}                                        
      state: ""                                           
    name: restore                                         
    object:                                               
      apiVersion: ""                                      
      group: ""                                           
      kind: statefulset                                   
      name: mysql                                         
      namespace: mysql                                    
      resource: ""                                        
    phases:                                               
    - name: restorePhase1                                 
      progress:                                           
        lastTransitionTime: "2023-08-11T06:41:18Z"                                                                  
        progressPercent: "100"                            
      state: complete                                     
    - name: restorePhase2                                 
      progress: {}                                        
      state: pending                                      
    - name: restorePhase3                                 
      progress: {}                                        
      state: pending                                      
  error:                                                  
    message: ""                                           
  progress:                                               
    lastTransitionTime: "2023-08-11T06:41:18Z"            
    percentCompleted: "16"                                
    runningPhase: restorePhase1                           
  state: running
```

**3. First phase of both the action is completed and 2nd phase is running**

```jsx
status:                                                                                                                                                                                                                                 
  actions:                                                                                                                                                                                                                              
  - artifacts:                                                                                                                                                                                                                          
      operatorBackup:                                                                                                                                                                                                                   
        keyValue:                                                                                                                                                                                                                       
          path: '{{ .Phases.backupPhase3.Output.path }}'                                                                                                                                                                                
    blueprint: operator-blueprint                         
    deferPhase:                                           
      name: ""                                            
      progress: {}                                        
      state: ""                                           
    name: backup                                          
    object:                                               
      apiVersion: ""                                      
      group: ""                                           
      kind: statefulset                                   
      name: mysql                                         
      namespace: mysql                                    
      resource: ""                                        
    phases:                                               
    - name: backupPhase1                                  
      output:                                             
        path: /operator-backups/2023-08-11T06-41-03/backup.tgz                                                      
      progress:                                           
        lastTransitionTime: "2023-08-11T06:41:22Z"                                                                  
        progressPercent: "100"                            
      state: complete                                     
    - name: backupPhase2                                  
      progress: {}                                        
      state: running                                      
    - name: backupPhase3                                  
      progress: {}                                        
      state: pending                                      
  - blueprint: operator-blueprint                         
    deferPhase:                                           
      name: ""                                            
      progress: {}                                        
      state: ""                                           
    name: restore                                         
    object:                                               
      apiVersion: ""                                      
      group: ""                                           
      kind: statefulset                                   
      name: mysql                                         
      namespace: mysql                                    
      resource: ""                                        
    phases:                                               
    - name: restorePhase1                                 
      progress:                                           
        lastTransitionTime: "2023-08-11T06:41:18Z"                                                                  
        progressPercent: "100"                            
      state: complete                                     
    - name: restorePhase2                                 
      progress:                                           
        lastTransitionTime: "2023-08-11T06:41:23Z"                                                                  
        progressPercent: "0"                              
      state: running                                      
    - name: restorePhase3                                 
      progress: {}                                        
      state: pending                                      
  error:                                                  
    message: ""                                           
  progress:                                               
    lastTransitionTime: "2023-08-11T06:41:22Z"            
    percentCompleted: "33"                                
    runningPhase: backupPhase2                            
  state: running
```

**4. 2nd phase of second action is completed**

```jsx
status:                                                                                                                                                                                                                       [849/4603]
  actions:                                                                                                                                                                                                                              
  - artifacts:                                                                                                                                                                                                                          
      operatorBackup:                                                                                                                                                                                                                   
        keyValue:                                                                                                                                                                                                                       
          path: '{{ .Phases.backupPhase3.Output.path }}'                                                                                                                                                                                
    blueprint: operator-blueprint                                                                                                                                                                                                       
    deferPhase:                                                                                                                                                                                                                         
      name: ""                                                                                                                                                                                                                          
      progress: {}                                                                                                                                                                                                                      
      state: ""                                                                                                                                                                                                                         
    name: backup                                                                                                                                                                                                                        
    object:                                                                                                                                                                                                                             
      apiVersion: ""                                                                                                                                                                                                                    
      group: ""                                                                                                                                                                                                                         
      kind: statefulset                                                                                                                                                                                                                 
      name: mysql                                                                                                                                                                                                                       
      namespace: mysql                                                                                                                                                                                                                  
      resource: ""                                                                                                                                                                                                                      
    phases:                                               
    - name: backupPhase1                                                                                            
      output:                                             
        path: /operator-backups/2023-08-11T06-41-03/backup.tgz                                                      
      progress:                                           
        lastTransitionTime: "2023-08-11T06:41:22Z"                                                                  
        progressPercent: "100"                                                                                      
      state: complete                                     
    - name: backupPhase2                                  
      progress:                                                                                                     
        lastTransitionTime: "2023-08-11T06:41:27Z"                                                                  
        progressPercent: "0"                                                                                        
      state: running                                      
    - name: backupPhase3                                                                                            
      progress: {}                                                                                                  
      state: pending                                      
  - blueprint: operator-blueprint                         
    deferPhase:                                           
      name: ""                                                                                                      
      progress: {}                                        
      state: ""                                           
    name: restore                                         
    object:                                               
      apiVersion: ""                                      
      group: ""                                           
      kind: statefulset                                   
      name: mysql                                         
      namespace: mysql                                    
      resource: ""                                        
    phases:                                               
    - name: restorePhase1                                 
      progress:                                           
        lastTransitionTime: "2023-08-11T06:41:18Z"                                                                  
        progressPercent: "100"                            
      state: complete                                     
    - name: restorePhase2                                 
      progress:                                           
        lastTransitionTime: "2023-08-11T06:41:33Z"                                                                  
        progressPercent: "100"                            
      state: complete                                     
    - name: restorePhase3                                 
      progress: {}                                        
      state: running                                      
  error:                                                  
    message: ""                                           
  progress:                                               
    lastTransitionTime: "2023-08-11T06:41:33Z"            
    percentCompleted: "50"                                
    runningPhase: restorePhase3                           
  state: running
```

**5. All the actions are completed**

```jsx
status:                                                                                                             
  actions:                                                                                                          
  - artifacts:                                                                                                      
      operatorBackup:                                                                                               
        keyValue:                                                                                                   
          path: /operator-backups/2023-08-11T06-41-03/backup.tgz                                                    
    blueprint: operator-blueprint                         
    deferPhase:                                           
      name: ""                                            
      progress: {}                                        
      state: ""                                           
    name: backup                                          
    object:                                                                                                         
      apiVersion: ""                                                                                                
      group: ""                                                                                                     
      kind: statefulset                                                                                             
      name: mysql                                                                                                   
      namespace: mysql                                                                                              
      resource: ""                                                                                                  
    phases:                                                                                                         
    - name: backupPhase1                                                                                            
      output:                                                                                                       
        path: /operator-backups/2023-08-11T06-41-03/backup.tgz                                                      
      progress:                                                                                                     
        lastTransitionTime: "2023-08-11T06:41:22Z"                                                                  
        progressPercent: "100"                                                                                      
      state: complete                                                                                               
    - name: backupPhase2                                                                                            
      output:                                                                                                       
        path: /operator-backups/2023-08-11T06-41-03/backup.tgz                                                      
      progress:                                                                                                     
        lastTransitionTime: "2023-08-11T06:41:42Z"                                                                  
        progressPercent: "100"                                                                                      
      state: complete                                                                                               
    - name: backupPhase3                                                                                            
      output:                                                                                                       
        path: /operator-backups/2023-08-11T06-41-03/backup.tgz                                                      
      progress:                                                                                                     
        lastTransitionTime: "2023-08-11T06:42:02Z"                                                                  
        progressPercent: "100"                            
      state: complete                                     
  - blueprint: operator-blueprint                         
    deferPhase:                                           
      name: ""                                            
      progress: {}                                        
      state: ""                                           
    name: restore                                         
    object:                                                                                                         
      apiVersion: ""                                                                                                
      group: ""                                                                                                     
      kind: statefulset                                                                                             
      name: mysql                                                                                                   
      namespace: mysql                                                                                              
      resource: ""                                                                                                  
    phases:                                                                                                         
    - name: restorePhase1                                                                                           
      progress:                                                                                                     
        lastTransitionTime: "2023-08-11T06:41:18Z"                                                                  
        progressPercent: "100"                                                                                      
      state: complete                                                                                               
    - name: restorePhase2                                                                                           
      progress:                                                                                                     
        lastTransitionTime: "2023-08-11T06:41:33Z"                                                                  
        progressPercent: "100"                                                                                      
      state: complete                                                                                               
    - name: restorePhase3                                                                                           
      progress:                                                                                                     
        lastTransitionTime: "2023-08-11T06:41:47Z"                                                                  
        progressPercent: "100"                            
      state: complete                                     
  error:                                                  
    message: ""                                           
  progress:                                               
    lastTransitionTime: "2023-08-11T06:41:52Z"            
    percentCompleted: "100"                               
  state: complete
```

### CODE REVIEW

File: pkg/controller/controller.go [hunk 1]
```diff
@@ -471,10 +463,21 @@ func (c *Controller) runAction(ctx context.Context, t *tomb.Tomb, as *crv1alpha1
...
+				defer doneProgressTrack()
+				go func() {
+					// progress update is computed on a best-effort basis.
+					// if it exits with error, we will just log it.
+					if err := progress.TrackActionsProgress(progressTrackCtx, c.crClient, as.GetName(), as.GetNamespace(), p); err != nil {
+						log.Error().WithError(err)
+					}
+				}()
 				output, err = p.Exec(ctx, *bp, action.Name, *tp)
+				doneProgressTrack()
```
Comment (viveksinghggits) [inline pkg/controller/controller.go:476]: do we need to call it again even after calling it from `defer` ?

File: pkg/function/backup_data.go [hunk 2]
```diff
@@ -64,13 +68,19 @@ func init() {
...
 
 func (*backupDataFunc) Name() string {
 	return BackupDataFuncName
 }
 
-func (*backupDataFunc) Exec(ctx context.Context, tp param.TemplateParams, args map[string]interface{}) (map[string]interface{}, error) {
+func (b *backupDataFunc) Exec(ctx context.Context, tp param.TemplateParams, args map[string]interface{}) (map[string]interface{}, error) {
+	// Set progress percent
+	b.progressPercent = progress.PercentProgressStarted
+	defer func() { b.progressPercent = progress.PercentProgressCompleted }()
```
Comment (viveksinghggits) [inline pkg/function/backup_data.go:82]: would we want to set the progress as completed even if the function execution failed?

File: pkg/controller/controller.go [hunk 3]
```diff
@@ -471,10 +463,21 @@ func (c *Controller) runAction(ctx context.Context, t *tomb.Tomb, as *crv1alpha1
 			var msg string
 			if err == nil {
 				c.updateActionSetRunningPhase(ctx, as, p.Name())
+				progressTrackCtx, doneProgressTrack := context.WithCancel(ctx)
+				defer doneProgressTrack()
+				go func() {
+					// progress update is computed on a best-effort basis.
+					// if it exits with error, we will just log it.
+					if err := progress.TrackActionsProgress(progressTrackCtx, c.crClient, as.GetName(), as.GetNamespace(), p); err != nil {
```
Comment (viveksinghggits) [inline pkg/controller/controller.go:471]: `TrackActionsProgress` continuously updates the current progress in actionset phase right?
Should we rename it to something that is more clearly conveying that, for example `IterateActionProgress`? Basically a word that conveys something is being improved continuously. :smile:

**Comment by viveksinghggits** [comment]
The other question that I had was, how are we going to handle the Kopia related function. Some functions are already, in so they are going to break right?

File: pkg/controller/controller.go (hunk 1 above)
Comment (PrasadG193) [reply pkg/controller/controller.go:476]: Yes, the `doneProgressTrack()` needs to be called as soon as `Exec()` is done. The `defer()` gets called once all the phase Exec is done and action is completed (or failed). When the next phase execution starts during interaction in this function, the older phase's Progress() starts sending completion percent as 0. So before running next phase, the goroutine for current phase needs to be closed. Let me know if this makes sense or not.

File: pkg/function/backup_data.go (hunk 2 above)
Comment (PrasadG193) [reply pkg/function/backup_data.go:82]: No, for failed phase, the progress will be nil. I've added the sample in the test plan
```
status:                                                                                                                                                                                                     
  actions:                                                                                                                                                                                                            
  - artifacts:                                                                                                                                                                                                        
      operatorBackup:                                                                                                                                                                                                 
        keyValue:                                                                                                                                                                                                     
          path: '{{ .Phases.phase3.Output.path }}'                                                                                                                                                                    
    blueprint: operator-blueprint                                                                                                                                                                                     
    deferPhase:                                                                                                                                                                                                       
      name: ""                                                                                                                                                                                                        
      progress: {}                                                                                                                                                                                                    
      state: ""                                                                                                                                                                                                       
    name: backup                                                                                                                                                                                                      
    object:                                                                                                                                                                                                           
      apiVersion: v1                                                                                                                                                                                                  
      group: apps                                                                                                                                                                                                     
      kind: statefulset                                                                                                                                                                                               
      name: mysql                                                                                                                                                                                                     
      namespace: mysql                                                                                                                                                                                                
      resource: statefulsets                                                                                                                                                                                          
    phases:                                                                                                                                                                                                           
    - name: phase1                                                                                                                                                                                                    
      output:                                                                                                                                                                                                         
        path: /operator-backups/2023-05-22T07-32-36/backup.tgz                                                                                                                                                        
      progress:                                                                                                                                                                                                       
        lastTransitionTime: "2023-05-22T07:32:54Z"                                                                                                                                                                    
        progressPercent: "100"                                                                                                                                                                                         
      state: complete                                                                                                                                                                                                 
    - name: phase2                                                                                                                                                                                                    
      progress: {}                                                                                                                                                                                                    
      state: failed                                                                                                                                                                                                   
    - name: phase3                                                                                                                                                                                                    
      progress: {}                                                                                                                                                                                                    
      state: pending                                                                                                                                                                                                  
  error:                                                                                                                                                                                                              
    message: "Failed while waiting for Pod kanister-job-sk2bk to complete: Pod failed                                                                                                                                 
      or did not transition into complete state: Running backup\
     .
     .
  progress:                                                                                                                                                             
    lastTransitionTime: "2023-05-22T07:32:54Z"                                          
    percentCompleted: "33"                                                           
  state: failed  
```

File: pkg/kanister.go [hunk 4]
```diff
@@ -39,6 +40,7 @@ type Func interface {
 	RequiredArgs() []string
 	Arguments() []string
 	Exec(context.Context, param.TemplateParams, map[string]interface{}) (map[string]interface{}, error)
+	ExecProgress() (crv1alpha1.PhaseProgress, error)
```
[the remark this answers is not in the document]
Comment (PrasadG193) [reply pkg/kanister.go:43]: Makes sense, will make the changes

File: pkg/controller/controller.go (hunk 3 above)
Comment (viveksinghggits) [reply pkg/controller/controller.go:471]: I can settle with `UpdateActionsetsProgress`

File: pkg/controller/controller.go [hunk 5]
```diff
@@ -471,10 +463,21 @@ func (c *Controller) runAction(ctx context.Context, t *tomb.Tomb, as *crv1alpha1
 			var msg string
 			if err == nil {
 				c.updateActionSetRunningPhase(ctx, as, p.Name())
+				progressTrackCtx, doneProgressTrack := context.WithCancel(ctx)
+				defer doneProgressTrack()
+				go func() {
+					// progress update is computed on a best-effort basis.
+					// if it exits with error, we will just log it.
+					if err := progress.UpdateActionSetsProgress(progressTrackCtx, c.crClient, as.GetName(), as.GetNamespace(), p); err != nil {
```
Comment (viveksinghggits) [inline pkg/controller/controller.go:471]: Just wanted to understand that, we update the progress only when we start executing a phase? It's not continuously running in background and updating the status right?

**Review by viveksinghggits — COMMENTED** [review]
looks good to me almost, just some small questions.

File: pkg/apis/cr/v1alpha1/types.go [hunk 6]
```diff
@@ -187,6 +188,27 @@ type Phase struct {
 	State State `json:"state"`
 	// Output is the map of output artifacts produced by the Blueprint phase.
 	Output map[string]interface{} `json:"output,omitempty"`
+	// Progress represents the phase execution progress.
+	Progress PhaseProgress `json:"progress,omitempty"`
```
[the remark this answers is not in the document]
Comment (PrasadG193) [reply pkg/apis/cr/v1alpha1/types.go:192]: @viveksinghggits I think there is some confusion, the `Progress` field above is added in `Phase` struct. ie.`ActionSetStatus.Action[0].Phase.Progress`, not at `ActionSetStatus.Progress`
e.g this is how the status looks like
```
status:
  actions:
  - artifacts:
      operatorBackup:
        keyValue:
          path: /operator-backups/2023-04-19T18-42-49/backup.tgz
    blueprint: operator-blueprint
    deferPhase:
      name: ""
      progress: {}
      state: ""
    name: backup
    object:
      apiVersion: ""
      group: ""
      kind: statefulset
      name: mysql
      namespace: mysql
      resource: ""
    phases:
    - name: phase1
      output:
        path: /operator-backups/2023-04-19T18-42-49/backup.tgz
      progress:
        lastTransitionTime: "2023-04-19T18:43:10Z"
        progressPercent: "100"
      state: complete
    - name: phase2
      output:
        path: /operator-backups/2023-04-19T18-42-49/backup.tgz
      progress:
        lastTransitionTime: "2023-04-19T18:43:28Z"
        progressPercent: "100"
      state: complete
    - name: phase3
      output:
        path: /operator-backups/2023-04-19T18-42-49/backup.tgz
      progress:
        lastTransitionTime: "2023-04-19T18:43:48Z"
        progressPercent: "100"
      state: complete
  error:
    message: ""
  progress:
    lastTransitionTime: "2023-04-19T18:43:48Z"
    percentCompleted: "100"
  state: complete
```

File: pkg/controller/controller.go (hunk 5 above)
Comment (PrasadG193) [reply pkg/controller/controller.go:471]: @viveksinghggits It is continuously running as a goroutine and checks for progress after every interval. And updates the status if there is any change in it

**Comment by PrasadG193** [comment]
@pavannd1 could you please take a look?

File: pkg/controller/controller.go [hunk 7]
```diff
@@ -491,8 +494,17 @@ func (c *Controller) runAction(ctx context.Context, t *tomb.Tomb, as *crv1alpha1
 				coreErr = nil
 				rf = func(ras *crv1alpha1.ActionSet) error {
 					ras.Status.Actions[aIDX].Phases[i].State = crv1alpha1.StateComplete
+					pp, err := p.Progress()
+					if err != nil {
+						log.Error().WithError(err)
+						return nil
+					}
+					ras.Status.Actions[aIDX].Phases[i].Progress = pp
```
Comment (pavannd1) [inline pkg/controller/controller.go:502]: Could this conflict with the update in `setActionSetPhaseProgress`?

File: pkg/controller/controller.go (hunk 5 above)
Comment (PrasadG193) [reply pkg/controller/controller.go:471]: @pavannd1 its shows nil progress if the phase is in pending state. Please check the samples I've added in the test plan.
E.g 
```
    phases:
    - name: phase1
      output:
        path: /operator-backups/2023-04-19T18-42-49/backup.tgz
      progress:
        lastTransitionTime: "2023-04-19T18:43:10Z"
        progressPercent: "100"
      state: complete
    - name: phase2
      progress: {}
      state: running
    - name: phase3
      progress: {}
      state: pending
  error:
    message: ""
  progress:
    lastTransitionTime: "2023-04-19T18:43:10Z"
    percentCompleted: "33"
    runningPhase: phase2
  state: running
```

File: pkg/apis/cr/v1alpha1/types.go [hunk 8]
```diff
@@ -187,6 +188,27 @@ type Phase struct {
...
+// PhaseProgress represents the execution state of the phase.
+type PhaseProgress struct {
+	// ProgressPercent represents the execution progress in percentage.
+	ProgressPercent string `json:"progressPercent,omitempty"`
+	// SizeUploadedB represents the size of data uploaded in Bytes till now as a part of phase execution.
+	// This field will be empty for phases which do not involve data movement.
+	SizeUploadedB int64 `json:"sizeUploadedB,omitempty"`
+	// EstimatedUploadSizeB represents the total estimated size of data in Bytes
+	// that will be uploaded during the phase execution.
+	EstimatedUploadSizeB int64 `json:"estinatedUploadSizeB,omitempty"`
```
[the remark this answers is not in the document]
Comment (PrasadG193) [reply pkg/apis/cr/v1alpha1/types.go:204]: Yes, this filed will be ommited 
```
    - name: phase1                                                                                                                                                                                                    
      output:                                                                                                                                                                                                         
        path: /operator-backups/2023-05-22T07-32-36/backup.tgz                                                                                                                                                        
      progress:                                                                                                                                                                                                       
        lastTransitionTime: "2023-05-22T07:32:54Z"                                                                                                                                                                    
        progressPercent: "100"                                                                                                                                                                                         
      state: complete             
```

File: pkg/controller/controller.go [hunk 9]
```diff
@@ -528,6 +540,11 @@ func (c *Controller) runAction(ctx context.Context, t *tomb.Tomb, as *crv1alpha1
 func (c *Controller) updateActionSetRunningPhase(ctx context.Context, as *crv1alpha1.ActionSet, phase string) {
 	err := reconcile.ActionSet(ctx, c.crClient.CrV1alpha1(), as.Namespace, as.Name, func(as *crv1alpha1.ActionSet) error {
 		as.Status.Progress.RunningPhase = phase
+		for i := 0; i < len(as.Status.Actions[0].Phases); i++ {
```
[the remark this answers is not in the document]
Comment (PrasadG193) [reply pkg/controller/controller.go]: We are using wrong index here, pushing the fix.
We wanted to update the phase.State of running action here

**Comment by PrasadG193** [comment]
Added test case scenario when executing multiple actions with the same blueprint in the PR's test plan. @pavannd1 @viveksinghggits PTAL

**Comment by viveksinghggits** [comment]
> Added test case scenario when executing multiple actions with the same blueprint in the PR's test plan. @pavannd1 @viveksinghggits PTAL

I had already looked into the PR so I am going to skip :green_apple: , but just wanted to make sure that the issue that we talked about where two actions are running in parallel is not going to cause any issues with `runningPhase` right?
If yes, do we have a ticket/issue somewhere to track that?

**Comment by PrasadG193** [comment]
@viveksinghggits I've filed - https://github.com/kanisterio/kanister/issues/2270

**Comment by PrasadG193** [comment]
@pavannd1 could you please re-review the PR?

**Review by pavannd1 — APPROVED** [review]
LGTM. Feel free to merge once CI passes.

**Pull request merged**